### PR TITLE
[CARBONDATA-2475] Support Modular Core for Materialized View DataMap for query matching and rewriting

### PR DESCRIPTION
**Pull request by ravipesala**
Currently carbon supports preaggregate datamap, which only supports preaggregate on single table. To improve it, we can add join capability by implementing Materialized View.

In carbon Materialized View, Modular Core provides the plan matcher and rewrites the query with MV query. 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java [hunk 1]
```diff
@@ -208,7 +208,7 @@ public DataMapCatalog getDataMapCatalog(DataMapProvider dataMapProvider, String
   }
 
   /**
-   * Initialize by reading all datamaps from store and re register it
+   * Intialize by reading all datamaps from store and re register it
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java]: typo

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java [hunk 2]
```diff
@@ -53,6 +53,22 @@ private DataMapStatusManager() {
     return storageProvider.getDataMapStatusDetails();
   }
 
+  /**
+   * Get enabled datamap status details
+   * @return
+   * @throws IOException
+   */
+  public static DataMapStatusDetail[] getEnabledDataMapStatusDetails() throws IOException {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java]: I think this is unused function, no need to add

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala [hunk 3]
```diff
@@ -0,0 +1,676 @@
+package org.apache.carbondata.mv.rewrite
```
Comment (jackylk) [inline datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: please add license.
license is missing in all testcase files

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/Tpcds_1_4_Suite.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+package org.apache.carbondata.mv.rewrite
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.test.util.PlanTest
+import org.scalatest.BeforeAndAfter
+
+import org.apache.carbondata.mv.testutil.Tpcds_1_4_Tables._
+
+class Tpcds_1_4_Suite extends PlanTest with BeforeAndAfter {
```
Comment (jackylk) [inline datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/Tpcds_1_4_Suite.scala]: Please remove this file as it is all commented out

File: datamap/mv/plan/src/test/scala/org/apache/carbondata/mv/plans/Tpcds_1_4_BenchmarkSuite.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import scala.util.{Failure, Success, Try}
+
+import org.apache.spark.sql.SparkSession
+import org.scalatest.BeforeAndAfter
+
+import org.apache.carbondata.mv.dsl._
+import org.apache.carbondata.mv.testutil.ModularPlanTest
+
+// scalastyle:off println
+class Tpcds_1_4_BenchmarkSuite extends ModularPlanTest with BeforeAndAfter {
```
Comment (jackylk) [inline datamap/mv/plan/src/test/scala/org/apache/carbondata/mv/plans/Tpcds_1_4_BenchmarkSuite.scala]: Please remove this file

File: integration/spark2/src/main/spark2.2/org/apache/spark/sql/hive/CarbonAnalyzer.scala [hunk 6]
```diff
@@ -20,15 +20,33 @@ import org.apache.spark.sql.SparkSession
...
+    case e: Exception =>
+      null
+  }
+
   override def execute(plan: LogicalPlan): LogicalPlan = {
     var logicalPlan = analyzer.execute(plan)
     logicalPlan = CarbonPreAggregateDataLoadingRules(sparkSession).apply(logicalPlan)
-    CarbonPreAggregateQueryRules(sparkSession).apply(logicalPlan)
+    logicalPlan = CarbonPreAggregateQueryRules(sparkSession).apply(logicalPlan)
+    // TODO Get the analyzer rules from registered datamap class and apply here.
```
Comment (jackylk) [inline integration/spark2/src/main/spark2.2/org/apache/spark/sql/hive/CarbonAnalyzer.scala]: This TODO can be removed?

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4896/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5852/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4696/

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java (hunk 2 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java]: It is used in `SummaryDatasetCatalog`

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala (hunk 3 above)
Comment (ravipesala) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: ok

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/Tpcds_1_4_Suite.scala (hunk 4 above)
Comment (ravipesala) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/Tpcds_1_4_Suite.scala]: ok

File: datamap/mv/plan/src/test/scala/org/apache/carbondata/mv/plans/Tpcds_1_4_BenchmarkSuite.scala (hunk 5 above)
Comment (ravipesala) [reply datamap/mv/plan/src/test/scala/org/apache/carbondata/mv/plans/Tpcds_1_4_BenchmarkSuite.scala]: ok

File: integration/spark2/src/main/spark2.2/org/apache/spark/sql/hive/CarbonAnalyzer.scala (hunk 6 above)
Comment (ravipesala) [reply integration/spark2/src/main/spark2.2/org/apache/spark/sql/hive/CarbonAnalyzer.scala]: ok

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java]: ok

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4897/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5853/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4699/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4898/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5854/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4700/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**